DeusEx/Classes/DXREntranceRando.uc

```diff
@@ -237,37 +201,122 @@ function bool ValidateConnections()
             }
         }
     }
     
     //Theoretically I should probably actually check to see if the maps match,
     //but this is fairly safe...
     return visitable == numMaps;
+}
+
+function int GetAllMapNames(out MapConnection mapdests[15])
+{
+    local int i, j, numMaps;
+    local bool found;
+    for (i=0;i<numXfers;i++)
+    {
+        found=false;
+        for(j=0; j<numMaps; j++) {
+            if( xfers[i].mapname == mapdests[j].mapname ) {
+                found=true;
+                break;
+            }
+        }
+        if (found == false)
+        {
+            mapdests[numMaps].mapname = xfers[i].mapname;
+            mapdests[numMaps].numDests = 0;
+            numMaps++;
+        }
+    }
+    return numMaps;
+}
+
+function MarkMapsConnected(out MapConnection mapdests[15], int numMaps, string map_a, string map_b)
+{
+    local int mapidx;
+    mapidx = FindMapDestinations(mapdests, numMaps, map_a);
+    if(mapidx == -1) {
+        err("failed MarkMapsConnected("$numMaps$", "$map_a$", "$map_b$") find A");
+        return;
+    }
+    AddDestination( mapdests[mapidx], map_b);
     
+    mapidx = FindMapDestinations(mapdests, numMaps, map_b);
+    if(mapidx == -1) {
+        err("failed MarkMapsConnected("$numMaps$", "$map_a$", "$map_b$") find B");
+        return;
+    }
+    AddDestination( mapdests[mapidx], map_a);
+}
+
+function int FindMapDestinations(MapConnection mapdests[15], int numMaps, string mapname)
+{
+    local int i;
+    for (i=0;i<numMaps;i++)
+    {
+        if (mapdests[i].mapname == mapname)
+        {
+            return i;
+        }
+    }
+    err("failed to FindMapDestinations for "$mapname);
+    return -1;
+}
+
+function AddDestination( out MapConnection map, string mapname )
+{
+    local int i;
+    for(i=0; i < map.numDests; i++) {
+        if( map.dest[i] == mapname ) {
+            return;
+        }
+    }
+    map.dest[map.numDests] = mapname;
+    map.numDests++;
+}
+
+function MarkMapDestinationsVisitible(MapConnection m, out string canvisit[15], out int visitable)
+{
+    local int k, p;
+    local bool canVisitDest;
+
+    for (k=0;k<m.numDests;k++)
+    {
+        //If any of those destinations are not in the places we can visit, add them
+        canVisitDest = False;
+        for (p=0;p<visitable;p++)
+        {
+            if ( canvisit[p]==m.dest[k] )
+            {
+                canVisitDest = True;
+            }
+        }
+        if (canVisitDest == False)
+        {
+            canvisit[visitable] = m.dest[k];
+            visitable++;
+        }
+    }
 }
 
 function GenerateConnections(int missionNum)
+{
+    local int attempts;
+    local bool isValid;
+    isValid = False;
+    for(attempts=0; attempts<100; attempts++)
+    {
+        _GenerateConnections(missionNum);
+        if(ValidateConnections())
+            return;
+    }
+    err("GenerateConnections("$missionNum$") failed after "$attempts$" attempts!");
```

Comment:
an error that tells you what happened is better than just a crash, especially because the game doesn't flush write the logs when it crashes, so you have no idea where it crashed